### PR TITLE
Vehicle occupancy v2

### PR DESCRIPTION
**Pull request by julyto**
I created a new version just for vehicle occupancy fot gtfs-rt.

### CODE REVIEW

File: internal/vehicleoccupancies_v2/api_test.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	c.Request = httptest.NewRequest("GET", "/vehicle_occupancies", nil)
+	w := httptest.NewRecorder()
+	engine.ServeHTTP(w, c.Request)
+	require.Equal(200, w.Code)
+	err = json.Unmarshal(w.Body.Bytes(), &response)
+	require.Nil(err)
+	assert.Len(response.VehicleOccupancies, 0)
+	assert.Empty(response.Error)
+
+	c.Request = httptest.NewRequest("GET", "/vehicle_occupancies?date=20210118", nil)
```
Comment (azime) [inline internal/vehicleoccupancies_v2/api_test.go:64]:
```suggestion
	c.Request = httptest.NewRequest("GET", "/vehicle_occupancies?date=20210118T101010", nil)
```
Datetime, No ?

File: internal/vehicleoccupancies_v2/context.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+			// Filter on vehiclejourney_code
+			foundVjCode := FoundVjCode(*vo, param.VehicleJourneyCodes)
+
+			//Fileter on datetime (default value Now)
+			if vo.DateTime.Before(param.Date) {
+				//println("****************************** DATE BEFORE")
+				continue
+			}
+			//println("****************************** DATE AFTER")
+			//println("****************************** ", foundStopCode, foundVjCode)
```
Comment (azime) [inline internal/vehicleoccupancies_v2/context.go:106]: To remove?

File: internal/vehicleoccupancies_v2/context.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+		for _, vo := range d.VehicleOccupancies {
+			// Filter on stop_code
+			foundStopCode := FoundStopCode(*vo, param.StopCodes)
+
+			// Filter on vehiclejourney_code
+			foundVjCode := FoundVjCode(*vo, param.VehicleJourneyCodes)
+
+			//Fileter on datetime (default value Now)
+			if vo.DateTime.Before(param.Date) {
+				//println("****************************** DATE BEFORE")
```
Comment (azime) [inline internal/vehicleoccupancies_v2/context.go:102]: To remove?

File: internal/vehicleoccupancies_v2/factory.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+	GetRereshTime() string
+}
+
+// Patern factory Vehicle occupancies
+func VehicleOccupancyFactory(type_vehicleoccupancy string) (IVehicleOccupancy, error) {
+	if type_vehicleoccupancy == string(connectors.Connector_GRFS_RT) {
+		return &VehicleOccupanciesGtfsRtContext{}, nil
+		/*} else if type_vehicleoccupancy == string(connectors.Connector_ODITI) {
+		return &VehicleOccupanciesOditiContext{}, nil*/
```
Comment (azime) [inline internal/vehicleoccupancies_v2/factory.go:41]: To remove?

File: internal/vehicleoccupancies_v2/factory_test.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	vehicleoccupancies, err := VehicleOccupancyFactory(string(connectors.Connector_GRFS_RT))
+	require.Nil(err)
+	_, ok := vehicleoccupancies.(*VehicleOccupanciesGtfsRtContext)
+	require.True(ok)
+	/*
+		// Create vehicle occupancies type of ODITI
+		vehicleoccupancies, err = VehicleOccupancyFactory(string(connectors.Connector_ODITI))
+		require.Nil(err)
+		_, ok = vehicleoccupancies.(*VehicleOccupanciesGtfsRtContext)
+		require.False(ok)*/
```
Comment (azime) [inline internal/vehicleoccupancies_v2/factory_test.go:25]: To remove?

File: internal/vehicleoccupancies_v2/factory_test.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+		vehicleoccupancies, err = VehicleOccupancyFactory(string(connectors.Connector_ODITI))
+		require.Nil(err)
+		_, ok = vehicleoccupancies.(*VehicleOccupanciesOditiContext)
+		require.True(ok)
+
+		// Create vehicle occupancies type of ODITI
+		vehicleoccupancies, err = VehicleOccupancyFactory(string(connectors.Connector_GRFS_RT))
+		require.Nil(err)
+		_, ok = vehicleoccupancies.(*VehicleOccupanciesOditiContext)
+		require.False(ok)*/
```
Comment (azime) [inline internal/vehicleoccupancies_v2/factory_test.go:44]: To remove?

File: internal/vehicleoccupancies_v2/context.go (hunk 2 above)
Comment (julyto) [reply internal/vehicleoccupancies_v2/context.go:106]: :+1:

File: internal/vehicleoccupancies_v2/context.go (hunk 3 above)
Comment (julyto) [reply internal/vehicleoccupancies_v2/context.go:102]: :+1:

File: internal/vehicleoccupancies_v2/factory.go (hunk 4 above)
Comment (julyto) [reply internal/vehicleoccupancies_v2/factory.go:41]: I will need it fot Oditi

File: internal/vehicleoccupancies_v2/factory_test.go (hunk 5 above)
Comment (julyto) [reply internal/vehicleoccupancies_v2/factory_test.go:25]: I will need it fot Oditi

File: internal/vehicleoccupancies_v2/factory_test.go (hunk 6 above)
Comment (julyto) [reply internal/vehicleoccupancies_v2/factory_test.go:44]: I will need it fot Oditi

File: internal/vehicleoccupancies_v2/api_test.go (hunk 1 above)
Comment (julyto) [reply internal/vehicleoccupancies_v2/api_test.go:64]: The API only uses a date of type "20060102"

**Review by azime — APPROVED** [review]

File: internal/vehicleoccupancies_v2/context.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+		// Implement filter on parameters
+		for _, vo := range d.VehicleOccupancies {
+			// Filter on stop_code
+			foundStopCode := FoundStopCode(*vo, param.StopCodes)
+
+			// Filter on vehiclejourney_code
+			foundVjCode := FoundVjCode(*vo, param.VehicleJourneyCodes)
+
+			//Fileter on datetime (default value Now)
```
Comment (kadhikari) [inline internal/vehicleoccupancies_v2/context.go:100]: To optimize the code this test can be done at the beginning.
We can do as below:
if vo.DateTime.Before(param.Date) contienue;
if not FoundStopCode(*vo, param.StopCodes) continue;
if not FoundVjCode(*vo, param.VehicleJourneyCodes) continue;

File: internal/vehicleoccupancies_v2/api.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+type VehicleOccupanciesResponse struct {
+	VehicleOccupancies []VehicleOccupancy `json:"vehicle_occupancies,omitempty"`
+	Error              string             `json:"error,omitempty"`
+}
+
+// Structures and functions to read files for vehicle_occupancies are here
+type VehicleOccupancy struct {
+	Id                 int       `json:"_"`
+	VehicleJourneyCode string    `json:"vehicle_journey_code"`
+	StopCode           string    `json:"stop_code"`
```
Comment (kadhikari) [inline internal/vehicleoccupancies_v2/api.go:22]: stop_code or stop_point_code ?

File: internal/vehicleoccupancies_v2/context.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+			if vo.DateTime.Before(param.Date) {
+				continue
+			}
+
+			if len(param.StopCodes) == 0 || len(param.VehicleJourneyCodes) == 0 {
```
[the remark this answers is not in the document]
Comment (julyto) [reply internal/vehicleoccupancies_v2/context.go:105]: :+1:

File: internal/vehicleoccupancies_v2/context.go (hunk 7 above)
Comment (julyto) [reply internal/vehicleoccupancies_v2/context.go:100]: :+1:

File: internal/vehicleoccupancies_v2/api.go (hunk 8 above)
Comment (julyto) [reply internal/vehicleoccupancies_v2/api.go:22]: :+1:

**Review by kadhikari — APPROVED** [review]

**Pull request merged**